R/method.R

```diff
@@ -60,46 +61,40 @@ as_signature <- function(signature) {
   if (!is.list(signature)) {
     signature <- list(signature)
   }
-  is_valid_signature <- vlapply(signature, function(x) inherits(x, "R7_class") | inherits(x, "character"))
-  if (all(is_valid_signature)) {
-    return(signature)
-  }
 
-  invalid_indexes <- which(!is_valid_signature)
-  invalid_classes <- vcapply(signature[!is_valid_signature], function(x) fmt_classes(class(x)))
-
-  stop(
-    "`signature` must be a list of <R7_class> or a <character>:\n",
-    paste0(collapse = "\n",
-      sprintf("- `signature[%s]`: is %s", invalid_indexes, invalid_classes)
-    ),
-  call. = FALSE)
+  for (i in seq_along(signature)) {
+    signature[[i]] <- as_class(signature[[i]], arg_name = sprintf("signature[[%i]]", i))
+  }
+  signature
 }
 
 method_impl <- function(generic, signature, ignore) {
   out <- .Call(method_, generic, signature, ignore)
-  if (is.null(out)) {
-    # If no R7 method is found, see if there are any S3 methods registered
-    if (inherits(generic, "R7_generic")) {
-      args <- generic@dispatch_args
-      generic <- generic@name
-    } else {
-      generic <- find_function_name(generic, topenv(environment(generic)))
-      args <- names(formals(generic))
-    }
-    args <- setdiff(args, "...")
+  if (!is.null(out)) {
+    return(out)
+  }
 
-    out <- getS3method(generic, signature[[1]][[1]], optional = TRUE)
+  # If no R7 method is found, see if there are any S3 methods registered
+  if (inherits(generic, "R7_generic")) {
+    args <- generic@dispatch_args
+    generic <- generic@name
+  } else {
+    args <- names(formals(generic))
+    generic <- find_function_name(generic, topenv(environment(generic)))
+  }
```

Comment:
There was a `args <- setdiff(args, "...")` here before, was removing that intentional?

Reply:
Good catch, thanks.